### PR TITLE
Allow for recursive `@compiled` without materialization

### PR DESCRIPTION
**Pull request by hameerabbasi**
- **Materialize only in outermost nested `@compiled`.**
- **`pre-commit autoupdate`**

### CODE REVIEW

File: src/finch/tensor.py [hunk 1]
```diff
@@ -154,14 +157,27 @@ def __mod__(self, other):
...
-        if self.ndim != 2 or other.ndim != 2:
-            raise ValueError(
-                f"Both tensors must be 2-dimensional, but are: {self.ndim=} and {other.ndim=}."
+    @compiled()
+    def __matmul__(self, other: "Tensor") -> "Tensor":
+        if self.ndim < 2 or other.ndim < 2:
+            raise NotImplementedError(
+                "`@` is not implemented for arrays with less than two dimensions."
             )
-        return tensordot(self, other, axes=((-1,), (-2,)))
```
Comment (mtsokol) [inline src/finch/tensor.py:164]: As I checked today, with Finch 1.0 we don't need this hack - `tensordot` is fine to use.

**Comment by mtsokol** [comment]
I'll review it once it's rebased.

File: src/finch/tensor.py (hunk 1 above)
Comment (hameerabbasi) [reply src/finch/tensor.py:164]: This is more general as well; supports tensors with ndim >= 2.

Comment (mtsokol) [reply src/finch/tensor.py:164]: I think we still want this change reverted.

Comment (hameerabbasi) [reply src/finch/tensor.py:164]: Do we know if tensordot supports ndim > 2? If not I’d prefer to keep it this way.

Comment (willow-ahrens) [reply src/finch/tensor.py:164]: tensordot supports ndim > 2

Comment (hameerabbasi) [reply src/finch/tensor.py:164]: Done in https://github.com/finch-tensor/finch-tensor-python/pull/83/commits/1f6513c7da4b1516c34e75484e912338eaa7c7b6.

File: src/finch/tensor.py [hunk 2]
```diff
@@ -154,12 +157,11 @@ def __mod__(self, other):
     def __pow__(self, other):
         return self._elemwise_op("^", other)
 
-    def __matmul__(self, other):
-        # TODO: Implement and use mul instead of tensordot
-        # https://github.com/FinchTensor/finch-tensor-python/pull/22#issuecomment-2007884763
-        if self.ndim != 2 or other.ndim != 2:
-            raise ValueError(
-                f"Both tensors must be 2-dimensional, but are: {self.ndim=} and {other.ndim=}."
+    @compiled()
```
Comment (mtsokol) [inline src/finch/tensor.py]: We don't need `@compiled` decorator here.

File: src/finch/tensor.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
     SparseList,
     sparse_formats_names,
 )
+from .compiled import compiled
```
Comment (mtsokol) [inline src/finch/tensor.py]: No need to import it.

File: src/finch/compiled.py [hunk 4]
```diff
@@ -2,10 +2,66 @@
 from functools import wraps
 
 from .julia import jl
-from .tensor import Tensor
```
Comment (mtsokol) [inline src/finch/compiled.py:5]: We can't import it top level anymore?

File: src/finch/compiled.py [hunk 5]
```diff
@@ -2,10 +2,66 @@
...
+def _to_lazy_tensor(x, /):
+    from .tensor import Tensor
+
+    if isinstance(x, Tensor) and not jl.isa(x._obj, jl.Finch.LazyTensor):
+        return Tensor(jl.Finch.LazyTensor(x._obj))
+
+    return x
+
+
+def compiled(opt=None, *, force_materialization=False, tag: int | None = None):
```
Comment (mtsokol) [inline src/finch/compiled.py:32]: As it's a new feature I would like to see a few tests for it:
- Test with a nested `@compiled` decorators.
- Test with `force_materialization=True` which checks that for all lazy inputs the output is computed.

**Review by mtsokol — CHANGES_REQUESTED** [review]

File: src/finch/compiled.py (hunk 4 above)
Comment (hameerabbasi) [reply src/finch/compiled.py:5]: It led to a recursive import; anyway it's undone in https://github.com/finch-tensor/finch-tensor-python/pull/83/commits/87f384c06d507d9c9024d880b347977fa7a14a5f.

File: src/finch/tensor.py (hunk 3 above)
Comment (hameerabbasi) [reply src/finch/tensor.py]: Fixed in https://github.com/finch-tensor/finch-tensor-python/pull/83/commits/87f384c06d507d9c9024d880b347977fa7a14a5f.

File: src/finch/compiled.py (hunk 5 above)
Comment (hameerabbasi) [reply src/finch/compiled.py:32]: Done in https://github.com/finch-tensor/finch-tensor-python/pull/83/commits/4d45756fa3ed78c2c8785a15c81ffeda94d5cac8.

File: src/finch/tensor.py (hunk 2 above)
Comment (hameerabbasi) [reply src/finch/tensor.py]: Done in https://github.com/finch-tensor/finch-tensor-python/pull/83/commits/87f384c06d507d9c9024d880b347977fa7a14a5f.

File: src/finch/compiled.py [hunk 6]
```diff
@@ -2,8 +2,58 @@
 from functools import wraps
 
 from .julia import jl
+from .typing import JuliaObj, Any
 from .tensor import Tensor
-from .typing import JuliaObj
+
+
+def _recurse(x, /, *, f):
```
Comment (mtsokol) [inline src/finch/compiled.py]: Let's put type annotations back in.

File: src/finch/compiled.py [hunk 7]
```diff
@@ -2,8 +2,58 @@
...
+    if isinstance(x, tuple | list):
+        return type(x)(_recurse(xi, f=f) for xi in x)
+    if isinstance(x, dict):
+        return {k: _recurse(v, f=f) for k, v in x}
+    return f(x)
+
+
+def _recurse_iter(x, /):
```
Comment (mtsokol) [inline src/finch/compiled.py]: Let's put type annotations back in.

File: src/finch/compiled.py [hunk 8]
```diff
@@ -2,8 +2,58 @@
...
+    if isinstance(x, tuple | list):
+        for xi in x:
+            yield from _recurse_iter(xi)
+    if isinstance(x, dict):
+        for xi in x.values():
+            yield from _recurse_iter(xi)
+    yield x
+
+
+def _to_lazy_tensor(x: Tensor | Any, /):
```
Comment (mtsokol) [inline src/finch/compiled.py]: Return type.

File: src/finch/compiled.py [hunk 9]
```diff
@@ -29,46 +79,13 @@ def set_optimizer(opt: AbstractScheduler) -> None:
...
-            result_tensor = Tensor(jl.Finch.compute(result._obj, **kwargs))
-            return result_tensor
-
-        return wrapper_func
-
-    return inner
-
-
-def lazy(tensor: Tensor) -> Tensor:
+def lazy(tensor: Tensor):
```
Comment (mtsokol) [inline src/finch/compiled.py]: Return type.

File: src/finch/compiled.py [hunk 10]
```diff
@@ -29,46 +79,13 @@ def set_optimizer(opt: AbstractScheduler) -> None:
...
     if tensor.is_computed():
         return Tensor(jl.Finch.LazyTensor(tensor._obj))
     return tensor
 
 
-def compute(
-    tensor: Tensor, *, opt: AbstractScheduler | None = None, tag: int = -1
-) -> Tensor:
+def compute(tensor: Tensor, *, opt: AbstractScheduler | None = None, tag: int = -1):
```
Comment (mtsokol) [inline src/finch/compiled.py]: Return type.

**Review by mtsokol — APPROVED** [review]
LGTM! Just typing nitpicks and it's ready.

**Pull request merged**